File: 251003/zdanovich/Publisher/src/main/java/by/zdanovich/Publisher/services/MarkService.java
package by.zdanovich.Publisher.services;

import by.zdanovich.Publisher.DTOs.Requests.MarkRequestDTO;
import by.zdanovich.Publisher.DTOs.Responses.MarkResponseDTO;
import by.zdanovich.Publisher.models.Issue;
import by.zdanovich.Publisher.models.Mark;
import by.zdanovich.Publisher.repositories.IssueRepository;
import by.zdanovich.Publisher.repositories.MarkRepository;
import by.zdanovich.Publisher.utils.exceptions.NotFoundException;
import by.zdanovich.Publisher.utils.mappers.MarkMapper;
import lombok.RequiredArgsConstructor;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
@RequiredArgsConstructor
public class MarkService {
    private final MarkRepository markRepository;
    private final IssueRepository issueRepository;
    private final MarkMapper markMapper;

    public MarkResponseDTO save(Mark mark, long issueId) {
        Issue issue = issueRepository.findById(issueId).orElseThrow(() -> new NotFoundException("Issue with such id does not exist"));
        issue.getMarks().add(mark);
        mark.getIssues().add(issue);
        return markMapper.toMarkResponse(markRepository.save(mark));
    }

    public MarkResponseDTO save(MarkRequestDTO markRequestDTO) {
        Mark mark = markMapper.toMark(markRequestDTO);
        return markMapper.toMarkResponse(markRepository.save(mark));
    }

    public List<MarkResponseDTO> findAll() {
        return markMapper.toMarkResponseList(markRepository.findAll());
    }

    @CacheEvict(value = "marks", key = "#id")
    public MarkResponseDTO findById(long id) {
        return markMapper.toMarkResponse(markRepository.findById(id)
                .orElseThrow(() -> new NotFoundException("Mark with such id does not exist")));
    }

    @CacheEvict(value = "marks", key = "#id")
    public void deleteById(long id) {
        if (!markRepository.existsById(id))
            throw new NotFoundException("Mark with such id not found");
        markRepository.deleteById(id);
    }

    @CacheEvict(value = "marks", key = "#markRequestDTO.id")
    public MarkResponseDTO update(MarkRequestDTO markRequestDTO) {
        Mark mark = markMapper.toMark(markRequestDTO);
        return markMapper.toMarkResponse(markRepository.save(mark));
    }

    public boolean existsByName(String name) {
        return markRepository.existsByName(name);
    }

}

File: 251003/Kukhatskavolets/microblog/publisher/src/main/kotlin/by/kukhatskavolets/publisher/services/MarkService.kt
package by.kukhatskavolets.publisher.services

import by.kukhatskavolets.publisher.dto.requests.MarkRequestTo
import by.kukhatskavolets.publisher.dto.responses.MarkResponseTo
import by.kukhatskavolets.publisher.mappers.toEntity
import by.kukhatskavolets.publisher.mappers.toResponse
import by.kukhatskavolets.publisher.repositories.MarkRepository
import by.kukhatskavolets.publisher.repositories.TweetRepository
import org.springframework.cache.annotation.CacheEvict
import org.springframework.cache.annotation.CachePut
import org.springframework.cache.annotation.Cacheable
import org.springframework.cache.annotation.Caching
import org.springframework.stereotype.Service

@Service
class MarkService(
    private val markRepository: MarkRepository,
    private val tweetRepository: TweetRepository,
) {
    @Caching(
        evict = [
            CacheEvict(value = ["marks"], key = "'all_marks'")
        ]
    )
    fun createMark(markRequestTo: MarkRequestTo): MarkResponseTo {
        val mark = markRequestTo.toEntity()
        val savedMark = markRepository.save(mark)
        return savedMark.toResponse()
    }

    @Cacheable(value = ["marks"], key = "#id")
    fun getMarkById(id: Long): MarkResponseTo {
        val mark = markRepository.findById(id).orElseThrow { NoSuchElementException() }
        return mark.toResponse()
    }

    @Cacheable(value = ["marks"], key = "'all_marks'")
    fun getAllMarks(): List<MarkResponseTo> =
        markRepository.findAll().map { it.toResponse() }

    @Caching(
        put = [CachePut(value = ["marks"], key = "#id")],
        evict = [CacheEvict(value = ["marks"], key = "'all_marks'")]
    )
    fun updateMark(id: Long, markRequestTo: MarkRequestTo): MarkResponseTo {
        val updatedMark = markRequestTo.toEntity().apply { this.id = id }
        return markRepository.save(updatedMark).toResponse()
    }

    @Caching(
        evict = [
            CacheEvict(value = ["marks"], key = "#id"),
            CacheEvict(value = ["marks"], key = "'all_marks'")
        ]
    )
    fun deleteMark(id: Long) {
        markRepository.findById(id).orElseThrow { NoSuchElementException() }
        markRepository.deleteById(id)
    }

    @Cacheable(value = ["marksByTweets"], key = "#tweetId")
    fun getMarksByTweetId(tweetId: Long): List<MarkResponseTo> {
        val tweet = tweetRepository.findById(tweetId).orElseThrow { NoSuchElementException() }
        return tweet.marks.map { it.toResponse() }
    }
}